src/rust/catnap/mod.rs

```diff
@@ -181,281 +174,266 @@ impl CatnapLibOS {
-            }
-            // Parse result.
-            match queue.connect(remote, yielder).await {
-                Ok(()) => {
-                    queue.commit();
-                    (qd, OperationResult::Connect)
-                },
+    /// Asynchronous code to establish a connection to a remote endpoint. This function returns a coroutine that runs
+    /// asynchronously to connect a queue and performs any necessary mult-queue operations at the libOS-level after
+    /// the connect succeeds or fails.
+    fn connect_coroutine(
+        &self,
+        qd: QDesc,
+        remote: SocketAddrV4,
+        yielder: Yielder,
+    ) -> Result<Pin<Box<Operation>>, Fail> {
+        let queue: CatnapQueue = self.get_queue(qd)?;
+        Ok(Box::pin(async move {
+            // Wait for connect operation to complete.
+            match queue.do_connect(remote, yielder).await {
+                Ok(()) => (qd, OperationResult::Connect),
                 Err(e) => {
-                    queue.rollback();
                     warn!("connect() failed (qd={:?}, error={:?})", qd, e.cause);
                     (qd, OperationResult::Failed(e))
                 },
             }
-        });
-        let task_id: String = format!("Catnap::connect for qd={:?}", qd);
-        self.insert_task_into_scheduler(task_id, qd, &yielder_handle, coroutine)
+        }))
     }
 
-    /// Closes a socket.
-    pub fn close(&mut self, qd: QDesc) -> Result<(), Fail> {
+    /// Synchronously closes a CatnapQueue and its underlying POSIX socket.
+    pub fn close(&self, qd: QDesc) -> Result<(), Fail> {
         #[cfg(feature = "profiler")]
         timer!("catnap::close");
         trace!("close() qd={:?}", qd);
         // Issue close operation.
-        let mut qtable: RefMut<IoQueueTable<CatnapQueue>> = self.qtable.borrow_mut();
-        match qtable.get_mut(&qd) {
-            Some(queue) => {
-                queue.prepare_close()?;
-                match queue.close() {
-                    Ok(()) => {
-                        queue.commit();
-                        queue.prepare_closed()?;
-                        queue.cancel_pending_ops(Fail::new(libc::ECANCELED, "This queue was closed"));
-                        queue.commit();
-                        qtable.free(&qd);
-                        Ok(())
-                    },
-                    Err(e) => {
-                        queue.abort();
-                        Err(e)
-                    },
-                }
-            },
-            None => Err(Fail::new(libc::EBADF, "invalid queue descriptor")),
-        }
+        self.get_queue(qd)?.close()?;
+        // Remove the queue from the queue table.
+        self.qtable.borrow_mut().free(&qd);
+        Ok(())
     }
 
-    /// Asynchronous close
-    pub fn async_close(&mut self, qd: QDesc) -> Result<QToken, Fail> {
+    /// Synchronous code to asynchronously close a queue. This function schedules the coroutine that asynchronously
```

Comment:
This comment is a bit confusing. Is there a way to make it clearer? (e.g., possibly removing/rewording the first statement)

Reply:
I dont see how to reword this different from what Irene wrote initially. Please suggest.

Reply:
The first statement can be removed.

Reply:
Not sure... It refers to the synchronous part of the asynchronous close call.

Reply:
As discussed offline, it's ok to leave this as is.